platforms/android/run_module_app/src/main.cc

```diff
@@ -0,0 +1,209 @@
+#include <android_native_app_glue.h>
+
+#include "absl/strings/str_split.h"
+#include "absl/strings/string_view.h"
+#include "iree/base/initializer.h"
+#include "iree/base/status.h"
+#include "iree/modules/hal/hal_module.h"
+#include "iree/tools/vm_util.h"
+#include "iree/vm/api.h"
+
+namespace iree {
+namespace {
+
+const char* kAppTag = "iree-run-module";
+#define LOGI(...) \
+  ((void)__android_log_print(ANDROID_LOG_INFO, kAppTag, __VA_ARGS__))
+#define LOGE(...) \
+  ((void)__android_log_print(ANDROID_LOG_ERROR, kAppTag, __VA_ARGS__))
+
+const char kModuleFileName[] = "module.vmfb";
+const char kEntryFunctionFileName[] = "entry_function.txt";
+const char kInputsFileName[] = "inputs.txt";
+const char kDriverFileName[] = "driver.txt";
+
+// A struct containing information regarding one IREE VM module invocation.
+struct IreeModuleInvocation {
+  std::string module;
+  std::string entry_function;
+  std::string inputs;
+  std::string driver;
+
+  operator bool() const {
+    return !module.empty() && !entry_function.empty() && !inputs.empty() &&
+           !driver.empty();
+  }
+};
+
+// A class for loading IREE module invocation information from Android apk asset
+// files.
+class ModuleLoader {
+ public:
+  explicit ModuleLoader(android_app* app) : app_context_(app) {}
+  ~ModuleLoader() = default;
+
+  IreeModuleInvocation LoadModuleInvocation() {
+    IreeModuleInvocation invocation = {};
+    invocation.module = ReadFileAsset(kModuleFileName);
+    invocation.entry_function = ReadFileAsset(kEntryFunctionFileName);
+    invocation.inputs = ReadFileAsset(kInputsFileName);
+    invocation.driver = ReadFileAsset(kDriverFileName);
+    return invocation;
+  }
+
+ private:
+  // Reads the given asset file and returns its contents. Returns an empty
+  // string on error.
+  std::string ReadFileAsset(const char* file_name) {
+    AAssetManager* asset_manager = app_context_->activity->assetManager;
+    AAsset* asset =
+        AAssetManager_open(asset_manager, file_name, AASSET_MODE_BUFFER);
+    if (!asset) {
+      LOGE("failed to open file '%s' in assets", kModuleFileName);
+      return {};
+    }
+
+    size_t size_in_bytes = AAsset_getLength(asset);
+    std::string contents;
+    contents.resize(size_in_bytes);
+
+    AAsset_read(asset, const_cast<char*>(contents.data()), size_in_bytes);
+    AAsset_close(asset);
+
+    return contents;
+  }
+
+  android_app* app_context_;
+};
+
+Status RunModule(const IreeModuleInvocation& invocation) {
+  IREE_RETURN_IF_ERROR(iree_hal_module_register_types())
+      << "registering HAL types";
+  iree_vm_instance_t* instance = nullptr;
+  IREE_RETURN_IF_ERROR(
+      iree_vm_instance_create(iree_allocator_system(), &instance))
+      << "creating instance";
+
+  iree_vm_module_t* input_module = nullptr;
+  IREE_RETURN_IF_ERROR(LoadBytecodeModule(invocation.module, &input_module));
+
+  iree_hal_device_t* device = nullptr;
+  IREE_RETURN_IF_ERROR(CreateDevice(invocation.driver, &device));
```

Comment:
I'm a purist and would prefer to use IREE's C api directly, instead of the C++ helpers in https://github.com/google/iree/blob/main/iree/tools/vm_util.h, but up to you...

Reply:
I'll just leave it as-is. We probably want to clean up some of the C++ wrappers as you said during the 1:1, but that's a separate task.